链接关闭过程中的并发 bug？

压测过程中偶尔会出现如下错误：

[2019-05-15 11:28:33:130] [JRaft-RPC-Processor-14] [ERROR] - com.alipay.sofa.jraft.rpc.impl.AbstractBoltClientService$1.onResponse(AbstractBoltClientService.java:176) - Fail to run RpcResponseClosure, the request is group_id: "group20"
server_id: "10.15.70.35:8958"
peer_id: "10.15.72.2:8958"
term: 1
prev_log_term: 1
prev_log_index: 1706046
committed_index: 1757056

java.util.ConcurrentModificationException: null
at java.util.ArrayDeque$DeqIterator.next(ArrayDeque.java:643) ~[?:1.8.0_45]
at com.alipay.sofa.jraft.core.Replicator.onError(Replicator.java:836) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.util.ThreadId.unlock(ThreadId.java:96) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.core.Replicator.waitMoreEntries(Replicator.java:1300) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.core.Replicator.sendEntries(Replicator.java:1362) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.core.Replicator.sendEntries(Replicator.java:1314) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.core.Replicator.onRpcReturned(Replicator.java:1112) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.core.Replicator$3.run(Replicator.java:633) ~[jraft-core-1.2.5.jar:?]
at com.alipay.sofa.jraft.rpc.impl.AbstractBoltClientService$1.onResponse(AbstractBoltClientService.java:174) ~[jraft-core-1.2.5.jar:?]
at com.alipay.remoting.rpc.RpcInvokeCallbackListener$CallbackTask.run(RpcInvokeCallbackListener.java:169) ~[bolt-1.5.3.jar:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/NonReentrantLock.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/NonReentrantLock.java
@@ -57,6 +57,10 @@ public final class NonReentrantLock extends AbstractQueuedSynchronizer implement
         return isHeldExclusively();
     }
 
+    public Thread getOwner() {
+        return owner;
+    }
+
     @Override
     public Condition newCondition() {
         return new ConditionObject();

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
@@ -22,6 +22,9 @@ import java.util.List;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.Lock;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 /**
  * Replicator id with lock.
  *
@@ -30,6 +33,8 @@ import java.util.concurrent.locks.Lock;
  * 2018-Mar-29 10:59:47 AM
  */
 public class ThreadId {
+
+    private static final Logger       LOG           = LoggerFactory.getLogger(ThreadId.class);
 
     private final Object              data;
     private volatile NonReentrantLock lock          = new NonReentrantLock();
@@ -85,6 +90,8 @@ public class ThreadId {
             return;
         }
         if (!theLock.isHeldByCurrentThread()) {
+            LOG.warn("Fail to unlock with {}, the lock is held by {} and current thread is {}.", this.data,
+                theLock.getOwner(), Thread.currentThread());
             return;
         }
         // calls all pending errors before unlock
@@ -104,12 +111,13 @@ public class ThreadId {
         }
     }
 
-    private void doUnlock(Lock theLock) {
+    private void doUnlock(final NonReentrantLock theLock) {
         if (theLock != null) {
             try {
                 theLock.unlock();
-            } catch (final Exception ignored) {
-                // ignored
+            } catch (final Exception e) {
+                LOG.warn("Fail to unlock with {}, the lock is held by {}, current thread is {}.", this.data,
+                    theLock.getOwner(), Thread.currentThread(), e);
             }
         }
     }
@@ -144,8 +152,8 @@ public class ThreadId {
      *
      * @param errorCode error code
      */
-    public void setError(int errorCode) {
-        final Lock theLock = this.lock;
+    public void setError(final int errorCode) {
+        final NonReentrantLock theLock = this.lock;
         if (theLock == null) {
             return;
         }

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
@@ -116,7 +116,7 @@ public class ThreadId {
             try {
                 theLock.unlock();
             } catch (final Exception e) {
-                LOG.warn("Fail to unlock with {}, the lock is held by {}, current thread is {}.", this.data,
+                LOG.warn("Fail to unlock with {}, the lock is held by {} and current thread is {}.", this.data,
                     theLock.getOwner(), Thread.currentThread(), e);
             }
         }

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
@@ -67,12 +67,12 @@ public class ThreadId {
                 }
             }
         } catch (final InterruptedException e) {
-            Thread.currentThread().interrupt(); //reset
+            Thread.currentThread().interrupt(); // reset
             return null;
         }
-        //Got the lock, double checking state.
+        // Got the lock, double checking state.
         if (this.lock == null) {
-            //should release lock
+            // should release lock
             theLock.unlock();
             return null;
         }
@@ -80,8 +80,11 @@ public class ThreadId {
     }
 
     public void unlock() {
-        final Lock theLock = this.lock;
+        final NonReentrantLock theLock = this.lock;
         if (theLock == null) {
+            return;
+        }
+        if (!theLock.isHeldByCurrentThread()) {
             return;
         }
         // calls all pending errors before unlock
@@ -105,8 +108,8 @@ public class ThreadId {
         if (theLock != null) {
             try {
                 theLock.unlock();
-            } catch (final Exception e) {
-                //ignore
+            } catch (final Exception ignored) {
+                // ignored
             }
         }
     }

